### PR TITLE
Sample and reject for random scalars

### PR DESCRIPTION
**Pull request by mstraka100**
Implements sample and reject for random scalar generation

Tested in the prover service repo by generating and verifying 1000 proofs using a small toy circuit

### CODE REVIEW

File: src/groth16.cpp [hunk 1]
```diff
@@ -288,8 +288,30 @@ Prover<Engine>::prove(typename Engine::FrElement* wtns)
     typename Engine::FrElement s;
     typename Engine::FrElement rs;
 
-    E.fr.copy(r, E.fr.zero());
-    E.fr.copy(s, E.fr.zero());
+    //E.fr.copy(r, E.fr.zero());
```
Comment (ziaptos) [inline src/groth16.cpp:291]: We can delete these commented out lines.

File: src/groth16.cpp [hunk 2]
```diff
@@ -288,8 +288,30 @@ Prover<Engine>::prove(typename Engine::FrElement* wtns)
...
+    }
+
+    for (int cmp = 0; cmp >= 0;)
+    {
+        randombytes_buf(&s, sizeof(s));
+        s.v[3] &= 0x3FFFFFFFFFFFFFFFull;
+        auto s_copy      = s.v;
+        auto fr_mod_copy = fr_modulus;
+        cmp              = Fr_rawCmp(s_copy, fr_mod_copy);
+    }
```
Comment (ziaptos) [inline src/groth16.cpp:314]: This added code actually is supposed to replace the 4 code lines below (318,319,322,323). I guess we need to test again after removing the 4 code lines below.

**Review by ziaptos — CHANGES_REQUESTED** [review]

File: src/groth16.cpp [hunk 3]
```diff
@@ -288,8 +288,30 @@ Prover<Engine>::prove(typename Engine::FrElement* wtns)
     typename Engine::FrElement s;
     typename Engine::FrElement rs;
 
-    E.fr.copy(r, E.fr.zero());
-    E.fr.copy(s, E.fr.zero());
+    //E.fr.copy(r, E.fr.zero());
+    //E.fr.copy(s, E.fr.zero());
+
+
+    FrRawElement fr_modulus = {0x43E1F593F0000001ull, 0x2833E84879B97091ull,
```
Comment (alinush) [inline src/groth16.cpp:295]: Where did you get these constants from and can you somehow assert against a library-defined constant that this is the right modulus in the right endianness?

Comment (ziaptos) [reply src/groth16.cpp:295]: @alinush The number was provided by @mstraka100, I believe it was from the Groth Paper (@mstraka100 can you confirm). I did make sure it is encoded in the right endianness:

1. The fact that we used to have last byte set to zero suggested little endian
2. I double confirmed it by testing Fr_rawCmp with 0x1000 vs 0x0100 vs 0x0001, etc.

Comment (mstraka100) [reply src/groth16.cpp:295]: It was taken from the arkworks BN128 implementation - I will add a comment stating this and cross-check it with other codebases/documentation to be sure

File: src/groth16.cpp (hunk 2 above)
Comment (mstraka100) [reply src/groth16.cpp:315]: Removed the code and tested successfully - 1000 successful proofs in a row

File: src/groth16.cpp (hunk 3 above)
Comment (mstraka100) [reply src/groth16.cpp:295]: Added a comment

**Review by ziaptos — APPROVED** [review]
lgtm

**Pull request merged**